core/src/apps/base.py

```diff
@@ -114,11 +125,42 @@ async def handle_Ping(ctx: wire.Context, msg: Ping) -> Success:
     return Success(message=msg.message)
 
 
+async def handle_DoPreauthorized(
+    ctx: wire.Context, msg: DoPreauthorized
+) -> protobuf.MessageType:
+    authorization = storage.cache.get(
+        storage.cache.APP_BASE_AUTHORIZATION
+    )  # type: Authorization
+    if not authorization:
+        raise wire.ProcessError("No preauthorized operation")
+
+    req = await ctx.call_any(
+        PreauthorizedRequest(), *authorization.expected_wire_types()
+    )
+
+    handler = wire.find_registered_workflow_handler(ctx.iface, req.MESSAGE_WIRE_TYPE)
+    if handler is None:
+        return wire.unexpected_message()
+
+    workflow.close_others()
```

Comment:
I think the reason why I put it here is because if the device is soft-locked, a message is being processed and the user touches the screen, then the message's workflow will get closed. So I put this here to make sure the lockscreen, or anything else for that matter, gets closed first. But maybe it's too strict. We certainly want to do it for `SignTx`, mainly because it has animations, but also because it can take a long time. We might not need it in case of `GetOwnershipProof` or maybe we could only call it in soft-locked state?

Reply:
https://github.com/trezor/trezor-firmware/pull/1127/commits/099d007bff8b8dbd4a8e4df525adbdb8dd594e64 https://github.com/trezor/trezor-firmware/pull/1127/commits/9a4288ba04e67478c04ff9770a396e4c9680a032 I guess this should suffice.

Reply:
Reviewing this, I actually think `sign_tx` should call this unconditionally. So far it hasn't been a problem because the homescreen is noninteractive, but the initial progress bars (before first user-facing dialog) are also drawn "blind".

in fact it should be part of `progress.init()`, right?

Reply:
Maybe `report_init()`?

Reply:
Could be, yes. I think it doesn't make a difference, as `report_init` is called after calling a layout, so we are pretty much guaranteed that `close_others` is a no-op

Reply:
https://github.com/trezor/trezor-firmware/pull/1127/commits/0863dfdfa1b87d36f8dcd3e11cfd162aa72801ff